### PR TITLE
Fix ext-ds object diffing and merging

### PR DESCRIPTION
**Pull request by simPod**
_I have not found a  way to test this on my project yet._

Rather than creating issue I tried to fix stubs or at least show the idea. The values do not need to have the same type and this should address it.

### CODE REVIEW

File: stubs/ext-ds.stub [hunk 1]
```diff
@@ -593,8 +599,9 @@ final class Set implements ArrayAccess, Collection, Traversable
 	}
 
 	/**
-	 * @param Set<TValue> $set
-	 * @return Set<TValue>
+	 * @template TValue2
+	 * @param Set<TValue2> $set
+	 * @return Set<TValue|TValue2>
```
Comment (enumag) [inline stubs/ext-ds.stub]: Actually this is wrong. `diff` only returns values from the first set that are not in the second set. So the result won't contain `TValue2`. Same goes for `Map::diff()`

File: stubs/ext-ds.stub [hunk 2]
```diff
@@ -153,16 +154,18 @@ final class Map implements IteratorAggregate, ArrayAccess, Collection
 	}
 
 	/**
-	 * @param Map<TKey, TValue> $map
-	 * @return Map<TKey, TValue>
+	 * @template TValue2
+	 * @param Map<TKey, TValue2> $map
+	 * @return Map<TKey, TValue|TValue2>
```
Comment (enumag) [inline stubs/ext-ds.stub]: This is wrong too. Since this is intersect the result values have to be `TValue&TValue2`

**Comment by simPod** [comment]
I'll write some tests to make the behavior 100% right

**Comment by enumag** [comment]
Tests aren't necessary in my opinion. What do you think @ondrejmirtes?

**Comment by ondrejmirtes** [comment]
You can write some using NodeScopeResolverTest::testFileAsserts() if you want to, it can’t hurt.

File: stubs/ext-ds.stub (hunk 2 above)
Comment (simPod) [reply stubs/ext-ds.stub]: Actually should stay as `TValue`
```php
     * @return Map A new map containing the pairs of the current instance
     *                 whose keys are also present in the given map. In other
     *                 words, returns a copy of the current map with all keys
     *                 removed that are not also in the other map.
     */
```

**Comment by simPod** [comment]
To sum up:

`diff()` and `intersect()` only modify current map based on keys  
`merge()`, `xor()`, `union()` result in some mixed kind of map C from maps A and B

I think that that enumerates all methods that need to be modified here.

**Comment by simPod** [comment]
I agree `TValue2` sucks a bit. Looked around for some inspiration but found none...

**Comment by enumag** [comment]
Another thought for map: shouldn't there be a `TKey2` used as well? `Map::intersect()` would return `Map<TKey&TKey2, TValue>`

**Comment by simPod** [comment]
@enumag intersect afaik returns only values whose keys are in both maps. So the keys must be of the same type to return anything.

But for the second group there should indeed be `TKey2`  and the resulting type is `Map<TKey|TKey2, TValue|TValue2>`

( merge(), xor(), union() )

**Comment by enumag** [comment]
> `intersect afaik returns only values whose keys are in both maps`

yup that's why the keys in resulting map need to be both, meaning `TKey&TKey2`

**Comment by simPod** [comment]
But both is union `|`, not intersection `&`, hm?

**Comment by simPod** [comment]
Also 

```php
$map = new Map([1=>1]);
$map->put('2', '2');
```

modifies the type but I have no idea how to implement that. Maybe through some extension?

Not sure why tests fail, can't spot a bug. Any idea?

**Comment by enumag** [comment]
> But both is union |, not intersection &, hm?

For the `intersect` method result to contain a key it needs to be in both maps. Meaning it needs to be of both type TKey and type TKey2 at the same time. Hence `&`.

**Comment by enumag** [comment]
As for https://github.com/phpstan/phpstan-src/pull/95#issuecomment-573322021 you should annotate such map with `/** @var Map<string|int, string|int>`.

**Comment by simPod** [comment]
> you should annotate such map with /** @var Map<string|int, string|int>.

not having to do that is desired, I'm trying to avoid all comments or unnecessary assertions in code if possible. 

It's possible with array but I'm not sure how that's implemented.

```php
<?php declare(strict_types = 1);

$array = ['a']; // initially array<string>
$array[] = 1; // phpstan now knows it's array<string|int>

foreach($array as $entry) {
	if(is_string($entry) || is_int($entry)) {
		continue;
	}
	
	throw new \Exception();
}
```
https://phpstan.org/r/f8a9f24b-de0b-424c-93da-02dafa902560

File: stubs/ext-ds.stub [hunk 3]
```diff
@@ -625,7 +637,8 @@ final class Set implements ArrayAccess, Collection, Traversable
 	}
 
 	/**
-	 * @param Set<TValue> $set
+	 * @template TValue2
+	 * @param Set<TValue2> $set
 	 * @return Set<TValue>
```
Comment (enumag) [inline stubs/ext-ds.stub]:
```suggestion
	 * @return Set<TValue&TValue2>
```

**Comment by enumag** [comment]
@simPod I don't know how that's done either. @ondrejmirtes ?

**Comment by ondrejmirtes** [comment]
@enumag Can you formulate your question? I don't know what you're asking about. This code seems OK: https://phpstan.org/r/f8a9f24b-de0b-424c-93da-02dafa902560

**Comment by enumag** [comment]
@ondrejmirtes https://github.com/phpstan/phpstan-src/pull/95#issuecomment-573411273

**Comment by ondrejmirtes** [comment]
Is the question how does this piece of code modifies the original object?

```php
$map = new Map([1=>1]);
$map->put('2', '2');
```

How is the object modified? What's the result of this following code?

```php
$map = new Map([1=>1]);
$map === 'aaa';
$map->put('2', '2');
$map === 'aaa';
```

**Comment by simPod** [comment]
I got lost a bit but the question is how to make this work:

```php
<?php declare(strict_types = 1);

$set = new Set('a'); // initially Set<string>
$set->add(1); // phpstan now does not know it's Set<string|int>. How to make PHPStan know?
```

**Comment by ondrejmirtes** [comment]
The actual dilema is:

1) Do you want to report it as a problem?
2) Or do you want to modify the type of the set?

I think it makes sense to do only one, and right now it's doing 1). Which is what you want from generics...

**Comment by simPod** [comment]
It is reported as a problem now and I'd like to modify the type of the set instead. Same behaviour as array has.

**Comment by ondrejmirtes** [comment]
If you want to make it `Set<string|int>` from the beginning you should use `@var`.

**Comment by ondrejmirtes** [comment]
I don't think this is how generics should work. When people write `Set<string>` they usually want to enforce that.

**Comment by ondrejmirtes** [comment]
Otherwise you should use mixed here and not TValue: https://github.com/phpstan/phpstan-src/blob/master/stubs/ext-ds.stub#L567-L572

**Comment by simPod** [comment]
> When people write Set<string> they usually want to enforce that.

Sure but what if I don't write it?  
It is still the same with arrays. If I declare `array<string>` I definitely want to enforce it. But if I write `$arr = [];` the inner type is not defined. Then this is possible: `$arr[] = 1; $arr[] = '1'`.

```php
// this should be possible
$set = new Set();
$set->add(1, '1');

// this is error
/** @var Set<string> $set */
$set = new Set();
$set->add(1, '1');
```

**Comment by ondrejmirtes** [comment]
I don't think this PR should be blocked on this request. Can I review it?

**Comment by simPod** [comment]
Definitely agree, I'll just address this https://github.com/phpstan/phpstan-src/pull/95#issuecomment-578422552 in 15m

**Comment by simPod** [comment]
Ok, I have not touched this https://github.com/phpstan/phpstan-src/blob/master/stubs/ext-ds.stub#L567-L572 so we can discuss it later.

Same would then be with eg. https://github.com/phpstan/phpstan-src/blob/master/stubs/ext-ds.stub#L227-L233.

You can review, thanks.

**Comment by ondrejmirtes** [comment]
The test needs to be skipped if `Ds` extension isn't loaded. (You can for example return an empty array from the data provider.) Otherwise it's good to merge 👍

**Comment by ondrejmirtes** [comment]
Next step should probably be something that would detect wrong usage of `Map::intersect()` and `Set::intersect()`. Some type intersections (like `string&int`) end up as `NeverType` and with the current stubs they wouldn't get detected.

It might be interesting to do this generally as part of generics implementation.

**Comment by ondrejmirtes** [comment]
Thank you!

**Comment by ondrejmirtes** [comment]
Could you maybe send some similar stubs to https://github.com/JetBrains/phpstorm-stubs? Then PHPStan could analyze the code even without the ds extension loaded. They're probably not intereted in generics since PhpStorm doesn't support them, but even with the usual `@param mixed` etc. it would be beneficial (not just for PHPStan).

**Comment by enumag** [comment]
Yeah I was planning to send them there because right now PHPStorm doesn't know about Ds classes at all so I have to use `composer install --dev php-ds/php-ds` for PHPStorm to even know that these classes exist. I was just waiting for this to be merged. I'll try to do it tomorrow.

**Comment by enumag** [comment]
See https://github.com/JetBrains/phpstorm-stubs/pull/741

**Pull request merged**